[WIP] Enemy HP bar

jroweboy opened the pull request:
Based off the UI changes I've made, this change adds an Enemy HP Bar for the last attacked enemy.

The code is currently a spaghetti mess. Turns out its really complicated to add this, so it may not be worth reviewing until I get a change to clean it up. I'm opening a Draft Pull Request now since its getting close to being ready, and if you wanted a general idea of what the code will be like and have some time to look it over, I feel I might as well open the PR early then.

TODO List:
Features:
- [x] Base enemy HP bar for last attacked enemy.
- [ ] Auto hide on area change.
- [ ] Auto hide on enemy despawn
- [ ] Hide HP bar after enemy dies
- [ ] Hide HP bar after a certain time limit. (Maybe don't need?)

Bug fixes:
- [ ] Certain combinations of HP/Max HP are displayed incorrectly so the enemy HP sometimes partially "refills".
- [ ] HP above 255 may not work properly (needs more testing)
- [ ] Probably more when mattrick gets his hands on it.

shicks commented on src/js/preshuffle.s:
```diff
@@ -781,15 +781,19 @@ MaybeSetCheckpointActual:
+  .byte $82             ; Ey
+; clear out the experience count thats there initially
+; and make room for something like enemy name
+.repeat 11
```
What do you think about putting the number on the left and allowing the name to go more in the bottom-middle?  I feel like that gives a better separation between the player's stats and the enemy stats.

shicks commented on src/js/preshuffle.s:
```diff
@@ -781,15 +781,19 @@ MaybeSetCheckpointActual:
 .endrep
+  ; hold spots here for the enemy hp amount
+  .byte $20,$20,$20,$9d ; _ _ _ /
+  .byte $20,$20,$20,$20 ; _ _ _ _
```
Thoughts about _not_ displaying enemy's MaxHP?  It would simplify the code quite a bit and removes some extra clutter from the HUD.  If you attention to the first hit, you can see (more or less) what it was.

shicks commented on src/js/preshuffle.s:
```diff
@@ -1009,6 +1038,61 @@ Do16BitSubtractionForEXP:
+      jsr DisplayNumberInternal
+    pla
+    tay
+CheckMax:
```
Can you make this a local label?  (I think it's an @-prefix, but I don't remember exactly)

shicks commented on src/js/preshuffle.s:
```diff
@@ -1075,12 +1159,19 @@ Do16BitSubtractionForEXP:
-;;; NOTE: must finish before 35152
-FREE_UNTIL $9152
+    sty LastAttackedEnemyOffset
+    jmp UpdateEnemyHPDisplay
```
I don't think this runs on a vanilla tink - can we add this there, too?

shicks reviewed (commented):
Looks awesome, thanks for moving this forward!

jroweboy commented:
I've rebased the code and added loading enemy names.

https://user-images.githubusercontent.com/952515/175800641-10516c55-969c-4635-a150-1e70920f44d7.mp4
